### ISSUE TITLE
return cleanup function from before & beforeEach

### ISSUE THREAD

**Issue by andykais** [mochajs/mocha#3873]
**Is your feature request related to a problem or a nice-to-have?? Please describe.**
I see a [common](https://stackoverflow.com/questions/20584233/mocha-pass-variable-to-the-next-test) _imperative_ pattern when doing test setup and teardown.
```ts
describe('test with setup', () => {
  let server
  beforeEach(() => {
    server = new Server()
  })
  afterEach(() => {
    server.close()
  })
})
```
The frustration is initializing a variable in a before scope that needs to be accessed by a after hook. Variables declared in the describe block should be for use in test cases, and adding uninitialized setup/teardown variables in the scope is confusing and stateful.

**Describe the solution you'd like**
[react hooks](https://reactjs.org/docs/hooks-effect.html) has a great pattern that addresses a similar problem they had with `componentWillMount` and `componentWillUnmount`. I propose adding similar functionality to the `before` and `beforeEach` blocks to mocha.
```ts
describe('test with setup', () => {
  beforeEach(() => {
    const server = new Server()
    return () => server.close()
  })
})
```
This should work fine with blocks that return promises as well:
```ts
describe('test with async setup', () => {
  before(() => {
    return createSomeFilesAsync()
      .then(() => () => deleteSomeFilesAsync()) // this line is the teardown
})
```

In short, the proposal is to allow `before` and `beforeEach` to return a function, which will be executed at the same time as `after` and `afterEach`, respectively. In the case of a returned promise, just resolve the promise and if it resolves to a function, it is a cleanup step.

**Describe alternatives you've considered**
the alternative is the first example I gave.

I would be interested in working on a pr for this if others find it interesting (and if I can receive some guidance from a contributor)

**Comment by Bamieh** [comment]
This would be considered a breaking change as our current pattern * does * allow return statements in the `before` blocks.

If a Promise is returned in the `before`/`beforeEach` then mocha waits for that promise to be resolved before starting to execute the test cases.

I would vote -1 on implementing such a breaking change as it is breaking the pattern we follow in the test cases and other code blocks in the interfaces.

**Comment by andykais** [comment]
I see the reasoning there, but is it a breaking change? It is fairly easy to detect if a returned object is a promise or not.

```js
// in before/beforeEach handler
if (returned.then) {
  // handle as a promise
} else if (typeof returned === 'function') {
  // handle as cleanup function
} else {
  // invalid returned value
}
```

I can see how it muddles the api, but I do not see how anyone would need to refactor code as a result.

**Comment by JoshuaKGoldberg** [comment]
Per #5027 we're not looking to make any major changes. This would constitute a breaking change for some users and is generally a difference in how the hooks run. Going to have to wontfix this. Someone yell at me if there's a really strong reason to re-open though!
